mm/shm/shmdt.c

```diff
@@ -178,4 +164,55 @@ int shmdt(FAR const void *shmaddr)
   return ERROR;
 }
 
+int shmdt(FAR const void *shmaddr)
+{
+  FAR struct tcb_s *tcb;
+  FAR struct mm_map_entry_s *entry;
+  FAR struct task_group_s *group;
+  int shmid;
+  int ret;
+
+  /* Get the TCB and group containing our virtual memory allocator */
+
+  tcb = nxsched_self();
+  DEBUGASSERT(tcb && tcb->group);
+  group = tcb->group;
+
+  /* Get exclusive access to process' mm_map */
+
+  ret = mm_map_lock();
+  if (ret == OK)
+    {
+      /* Perform the reverse lookup to get the shmid corresponding to this
+       * shmaddr. The mapping is matched with just shmaddr == map->vaddr.
+       */
+
+      entry = mm_map_find(shmaddr, 1);
+      if (!entry || entry->vaddr != shmaddr)
+        {
+          ret = -EINVAL;
+          shmerr("ERROR: No region matching this virtual address: %p\n",
+                 shmaddr);
+
+          mm_map_unlock();
+          return -EINVAL;
+        }
+
+      shmid = entry->priv.i;
+
+      /* Indicate that there is no longer any mapping for this region. */
+
+      if (mm_map_remove(get_group_mm(group), entry) < 0)
+        {
+          shmerr("ERROR: mm_map_remove() failed\n");
+        }
+
+      mm_map_unlock();
+
+      ret = shmdt_priv(tcb->group, shmaddr, shmid);
```

Comment:
why call shmdt_priv here and munmap_shm?

Reply:
There are two ways to unmap the shm region by the user. sysv "shmdt" call (see "man shmdt"), and "unmap" call. 

- sysv "shmdt" call is always done during normal user process execution
- unmap call could be done during normal process execution (but normally you'd use the shmdt if using this interface)
- unmap call can be done during proceess exit, if user has left the memory unmapped, or process dies unexpectedly

shmdt_priv is the common function which does the actual work for all cases.

Reply:
Yes, but mm_map_remove at line 205 will trigger munmap_shm and then shmdt_priv. Line 212 call shmdt_priv again, is it double free here?

Reply:
No, mm_map_remove doesn't trigger munmap_shm. It is vice versa, munmap_shm calls mm_map_remove.

So, mm_map_remove is called from either munmap or shmdt, depending on which one is used to unmap the region.

Reply:
System-V shm driver (mm/shm/) have to naturally implement the system-v interfaces (shmget, shmat, shmdt etc...). Of course, when that interface creates memory mappings, we have then two ways of deleting the mappings (posix unmap & system-v shmdt).

We can try to clean up the system-v shm interface, but I'd like to rather bring in the posix shm interface next, which I find more useful and interesting.

Reply:
I think I can do this cleanup soon, if you like

Reply:
I added https://github.com/apache/nuttx/pull/8068 to clean it up

Reply:
Thanks!